### ISSUE TITLE
[glsl-in] / [wgsl-out] float() on bool or vec of bool doesn't produce usable result

### ISSUE THREAD

**Issue by magcius** [gfx-rs/naga#1050]
```glsl
float TevPerCompGT(float a, float b) {
    return float(a > b);
}

vec3 TevPerCompGT(vec3 a, vec3 b) {
    return vec3(greaterThan(a, b));
}
```

Produces WGSL:

```wgsl
fn TevPerCompGT(a: f32, b: f32) -> f32 {
    var a1: f32;
    var b1: f32;

    a1 = a;
    b1 = b;
    let _e5: f32 = a1;
    let _e6: f32 = b1;
    return f32((_e5 > _e6));
}

fn TevPerCompGT1(a2: vec3<f32>, b2: vec3<f32>) -> vec3<f32> {
    var a3: vec3<f32>;
    var b3: vec3<f32>;

    a3 = a2;
    b3 = b2;
    let _e5: vec3<f32> = a3;
    let _e6: vec3<f32> = b3;
    return vec3<f32>((_e5 > _e6).xyz);
}
```

Tint complains: `Tint says: Parser: 456:12 error: cannot construct 'f32' with a value of type 'bool'`

**Comment by Gordon-F** [comment]
A slightly smaller example:
```glsl
#version 450
layout(location = 0) in vec2 inA;

void main() {
    float a = float(inA.x > inA.y);
}
```

Module:
```txt
expressions: {
    [1]: GlobalVariable(
        [1],
    ),
    [2]: Load {
        pointer: [1],
    },
    [3]: AccessIndex {
        base: [2],
        index: 0,
    },
    [4]: Load {
        pointer: [1],
    },
    [5]: AccessIndex {
        base: [4],
        index: 1,
    },
    [6]: Binary {
        op: Greater,
        left: [3],
        right: [5],
    },
    [7]: As {
        expr: [6],
        kind: Float,
        convert: Some(
            4,
        ),
    },
    [8]: LocalVariable(
        [1],
    ),
},
```

`glsl-in` should parse this case as `Expression::Select` instead of `Expression::Binary`.

**Comment by magcius** [comment]
I'm maybe more of the opinion that wgsl-out should transform `As([6], float)` to `Select(0.0f, 1.0f, [6])`, since other languages might be able to handle this cast natively. I don't know how you would output a component-wise Select from glsl-out, for instance.

**Comment by kvark** [comment]
Uh, it's also a question if float(bool) should be allowed by IR. It would be great to have a mini-investigation showing how various languages cope with this.

**Comment by magcius** [comment]
HLSL and GLSL both support it. Not sure about MSL.

**Comment by kvark** [comment]
how does it translate to/from SPIRV?

**Comment by magcius** [comment]
glslang seems to generate OpFOrdGreaterThan followed by OpSelect.

**Comment by kvark** [comment]
In cases like this I look at Rust, and it tells me:
> casting `bool` as `f32` is invalid

So let's validate that in the IR and not allow boolean casting to anything

**Comment by jakobhellermann** [comment]
I think this use case of casting booleans (and especially vec's) to floats is done way more often in shader languages as it is in rust.
Right now the smallest snippet to convert a vec3<bool> to a vec3<f32> is
```rust
fn vec3_bool_to_f32(bvec: vec3<bool>) -> vec3<f32> {
    let x = f32(i32(bvec.x));
	var y: f32 = 0.0;
	if(bvec.y) { y = 1.0; }
	var z: f32 = 0.0;
	if(bvec.z) { z = 1.0; }
    return vec3<f32>(x, y, z);
}
```
which is obviously not graet.

**Comment by kvark** [comment]
Small update from the upstream WGSL. This has been discussed, and we agreed to provide all of the things: bool -> scalar, scalar -> bool, and their vector variants.

**Comment by magcius** [comment]
> Right now the smallest snippet to convert a vec3 to a vec3 is

I believe you can use per-component select `select(vec3<f32>(1.0), vec3<f32>(0.0), my_bvec3);`

**Comment by magcius** [comment]
The glsl-in case was fixed in #1059 to use per-component select btw, even if WGSL is going to natively have this going forward. Closing this bug.
